### PR TITLE
ddl: fix alter int to bit will panic in backfill index

### PR DESCRIPTION
**Pull request by AilinKid**
Signed-off-by: AilinKid <314806019@qq.com>

### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/24253

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:
Root cause:
When changing type A to type B, in the write reorg state, we will build a row decoder to analyze the raw data.
For the example in the issue.

| A | B | changing column A|
|  ----  | ----  | ---- |
| int | int | bit(5) |

|decoder.mutRow type|
| ---- |
|int int int|

for the decoder.mutRow, the TiDB set its fieldType as `int int int` rather than `int int bit(5）`

In the reorg second stage, we will fetch those reorganized rows to build the index again.
when analyzing an old row like `19,1,19(bit(5))`, it will set `19(bit(15))` datum to a mutRow with int type, which will panic here. 

Thus we change changing column to use it's target type as below:
|decoder.mutRow type|
| ---- |
|int int bit(5)|

Moreover, under this mechanism, when changing int to char(10), we build the mutRow type like:
|decoder.mutRow type|
| ---- |
|int int char(10)|

In the reorg first stage, we will fetch those old rows to rebuild the row again. it will try to set `int datum` to mut row with `char(10)`, which will also be leading panic here.

### Solution

As vividly illustrated above, we decided to set the decode.mutRow as the target field type that it has. But we don't analyze the changing column and generated column when we decode the row map. because:
- For changing column, we shouldn't decode it here, because we will do a unified cast operation latter.
- For generated column, we didn't decode it here too, because the eval process may depend on the changing column.

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

- fix alter int to bit will panic in backfill index

### CODE REVIEW

File: util/rowDecoder/decoder.go [hunk 1]
```diff
@@ -175,3 +173,84 @@ func BuildFullDecodeColMap(cols []*table.Column, schema *expression.Schema) map[
...
+		if err != nil {
+			return nil, err
+		}
+		rd.mutRow.SetValue(colInfo.Offset, val.GetValue())
+	}
+	// return the existed column map here.
+	return row, nil
+}
+
+func (rd *RowDecoder) EvalRemainedExprColumnMap(ctx sessionctx.Context, sysLoc *time.Location, row map[int64]types.Datum) (map[int64]types.Datum, error) {
```
Comment (wjhuang2016) [inline util/rowDecoder/decoder.go:217]: Need comment.

File: ddl/column_type_change_test.go [hunk 2]
```diff
@@ -1752,3 +1752,27 @@ func (s *testColumnTypeChangeSuite) TestDDLExitWhenCancelMeetPanic(c *C) {
...
+		"  `a` int(11) DEFAULT NULL," +
+		"  `b` varchar(10) DEFAULT NULL," +
+		"  `c` decimal(10,2) DEFAULT NULL," +
+		"  KEY `idx1` (`a`)," +
+		"  UNIQUE KEY `idx2` (`a`)," +
+		"  KEY `idx3` (`a`,`b`)," +
+		"  KEY `idx4` (`a`,`b`,`c`)" +
+		") ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin")
+	tk.MustExec("insert into t values(19,1,1),(17,2,2)")
+	tk.MustExec("alter table t modify a bit(5) not null")
```
Comment (wjhuang2016) [inline ddl/column_type_change_test.go:1777]: Could you check the result?

File: util/rowDecoder/decoder.go [hunk 3]
```diff
@@ -104,10 +98,14 @@ func (rd *RowDecoder) DecodeAndEvalRowWithMap(ctx sessionctx.Context, handle kv.
 	for _, dCol := range rd.colMap {
 		colInfo := dCol.Col.ColumnInfo
 		val, ok := row[colInfo.ID]
-		if ok || dCol.GenExpr != nil {
+		if ok {
 			rd.mutRow.SetValue(colInfo.Offset, val.GetValue())
 			continue
 		}
+		if dCol.GenExpr != nil {
+			// skip for now, latter we will eval it.
```
Comment (wjhuang2016) [inline util/rowDecoder/decoder.go:106]: Where?

File: util/rowDecoder/decoder.go [hunk 4]
```diff
@@ -175,3 +173,84 @@ func BuildFullDecodeColMap(cols []*table.Column, schema *expression.Schema) map[
 func (rd *RowDecoder) CurrentRowWithDefaultVal() chunk.Row {
 	return rd.mutRow.ToRow()
 }
+
+// DecodeTheExistedColumnMap is used by ddl column-type-change first column reorg stage.
+// In the function, we only decode the existed column in the row and fill the default value.
+// For changing column, we shouldn't decode it here, because we will do a unified cast operation latter.
```
Comment (zimulala) [inline util/rowDecoder/decoder.go:179]: `we shouldn't decode it here` I think you mean don't cast.

File: util/rowDecoder/decoder.go [hunk 5]
```diff
@@ -175,3 +169,82 @@ func BuildFullDecodeColMap(cols []*table.Column, schema *expression.Schema) map[
...
+		}
+		rd.mutRow.SetValue(colInfo.Offset, val.GetValue())
+	}
+	// return the existed column map here.
+	return row, nil
+}
+
+// EvalRemainedExprColumnMap is used by ddl column-type-change first column reorg stage.
+// It is always called after DecodeTheExistedColumnMap to finish the generated column evaluation.
+func (rd *RowDecoder) EvalRemainedExprColumnMap(ctx sessionctx.Context, sysLoc *time.Location, row map[int64]types.Datum) (map[int64]types.Datum, error) {
```
Comment (zimulala) [inline util/rowDecoder/decoder.go:211]: Could we use this function instead of line122-158?

**Review by wjhuang2016 — APPROVED** [review]

**Pull request merged**